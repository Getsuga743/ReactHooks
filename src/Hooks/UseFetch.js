/* eslint-disable import/prefer-default-export */
import { useState, useEffect, useRef } from 'react';

export default (url) => {
  const isMounted = useRef(true);
  const [state, setState] = useState({ loading: true, error: null, data: null });
  useEffect(() => () => {
    isMounted.current = false;
  }, []);

  useEffect(() => {
    // -2
    setState({ loading: true, error: null, data: null });
    fetch(url).then((resp) => resp.json()).then((data) => {
      if (isMounted.current) {
        setState({
          loading: false,
          error: null,
          data,
        });
      }
    });
  }, [url]);
  return state;
};
// se ejecuta cuando la url cambia, el useEffect está "escuchando los cambios de la url"
// -2 esto hace q se resetee el objeto del UseState, permitiendo q se vuelva a poner el
//  div de loading en el Componente
// -3 usamos el useRef para trackear el estado del componente una vez q este esta montado,
// utilizamos el cuerpo del useEffect este comienza en true,
// utilizamos un useEffect, configuramos para q solo use el cleanup y
// que se llama una unica vez, la primera vez , al montarse, el cleanup cambia el estado del useRef
// en el segundo useEffect, hacemos que haya un useState condicional teniendo como condicion,
// la referencia de UseRef.
// De esta manera nos aseguramos de q el componente esté montado,
// cuando el useFetch asincrono es llamado.
// Esto sirve para evitar perdidas de memoria y crasheos en nuestra app
