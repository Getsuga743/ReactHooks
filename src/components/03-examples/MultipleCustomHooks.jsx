import React from 'react';
import { UseFetch } from '../../Hooks/UseFetch';
import { UseCounter } from '../../Hooks/UseCounter';

export default () => {
  const [countState, increment] = UseCounter(1);
  const { loading, error, data } = UseFetch(
    `https://www.breakingbadapi.com/api/quotes/${countState}`,
  );
  const { author, quote } = !!data && data[0];
  if (error) {
    console.log(error);
  }
  // -2
  return (
    <div>
      <h1>BreakingBad Quotes</h1>
      <hr />
      {
        // -1
        loading ? (
          <div className="alert alert-info textInfo">Loading...</div>
        ) : (
          <blockquote className="blockquote text-right">
            <p className="mb-0">{quote}</p>
            <footer className="blockquote-footer">{author}</footer>
            <button type="button" className="btn btn-primary mt-5" onClick={increment}>
              Next Quote
            </button>
          </blockquote>
        )
      }
    </div>
  );
};
// -1 utilizamos el operador ternario para hacer un renderizado condicional
// -2 si la data es null al negarla, pasa a un valor booleano de true, y al volver a complementarla
//  queda devuelta en false, pero ya podemos usarlo xq no es null
//  si data es true, entonces se muestra data[0]
//  mirar customHook useFetch
