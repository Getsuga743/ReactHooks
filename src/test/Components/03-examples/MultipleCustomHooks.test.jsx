/* eslint-disable no-undef */
import React from 'react';
import { shallow } from 'enzyme';
import MultipleCustomHooks from '../../../components/03-examples/MultipleCustomHooks';
import '../../../setupTest';
import UseFetch from '../../../Hooks/UseFetch';
// de esta forma le digo aq jest q ya se q lo que venga de useFetch es correcto y no me importa
jest.mock('../../../Hooks/UseFetch');

describe('pruebas en 03-examples, MultipleCustomHooks', () => {
  test('debe de mostrarse correctamente', () => {
    UseFetch.mockReturnValue({ data: null, loading: true, error: null });
    const wrapper = shallow(<MultipleCustomHooks />);
    expect(wrapper).toMatchSnapshot();
  });
  // aca se prueba q nuestro html del componente tenga lo mismo
  // que el resultado de nuestro mock del UseFetch
  test('debe de mostrar la informacion', () => {
    UseFetch.mockReturnValue({
      data: [{ author: 'fernando', quote: 'hola mundo' }],
      loading: false,
      error: null,
    });
    const wrapper = shallow(<MultipleCustomHooks />);
    console.log(wrapper.html());
    // nos aseguramos de q no exista la clase alert(...loading), osea q la informacion
    // se haya cargado, y que esta coincida con lo q le mandamos
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.alert').exists()).toBe(false);
    expect(wrapper.find('.mb-0').text().trim()).toBe('hola mundo');
  });
});
