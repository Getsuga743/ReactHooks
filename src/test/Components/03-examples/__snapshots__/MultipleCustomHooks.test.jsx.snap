// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en 03-examples, MultipleCustomHooks debe de mostrar la informacion 1`] = `
<div>
  <h1>
    BreakingBad Quotes
  </h1>
  <hr />
  <blockquote
    className="blockquote text-right"
  >
    <p
      className="mb-0"
    >
      hola mundo
    </p>
    <footer
      className="blockquote-footer"
    >
      fernando
    </footer>
    <button
      className="btn btn-primary mt-5"
      onClick={[Function]}
      type="button"
    >
      Next Quote
    </button>
  </blockquote>
</div>
`;

exports[`pruebas en 03-examples, MultipleCustomHooks debe de mostrarse correctamente 1`] = `
<div>
  <h1>
    BreakingBad Quotes
  </h1>
  <hr />
  <div
    className="alert alert-info textInfo"
  >
    Loading...
  </div>
</div>
`;
